Adding support for CloudStack capi images

davidjumani opened the pull request:
What this PR does / why we need it:

Adds support for CloudStack compatible CAPI images
This is created in anticipation of the repo migration request: https://github.com/kubernetes/org/issues/3279https://github.com/kubernetes/org/issues/3279

davidjumani commented:
Added membership requests
https://github.com/kubernetes/org/issues/3319
https://github.com/kubernetes/org/issues/3312

davidjumani commented:
/assign @codenrhoden

davidjumani commented:
@codenrhoden @CecileRobertMichon @EleanorRigby Can you please review this so it will be ready once the membership is added ? Thanks

CecileRobertMichon commented on images/capi/hack/image-convert.sh:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env bash
```
this file seems really specific to a specific use case / formats, can we name it something a little less generic than "image-convert.sh" ?

rohityadavcloud replied:
Agree @davidjumani, maybe rename this to convert-cloudstack-image.sh  or something more specific for CloudStack's use-case?

CecileRobertMichon commented:
/assign @rohityadavcloud

rohityadavcloud commented:
Thanks @CecileRobertMichon I'll review this (David and my org membership is pending here https://github.com/kubernetes/org/pull/3338 which is why prow bot can't assign it to me yet)

rohityadavcloud commented on OWNERS_ALIASES:
```diff
@@ -16,6 +16,9 @@ aliases:
     - jsturtevant
   image-builder-openstack-reviewers:
     - hidekazuna
+  image-builder-cloudstack-reviewers:
```
@davidjumani let's add @Pearl1594 here as well?

davidjumani replied:
From what's there in the file, there are at max 2 reviewers in each category. I've added Pearl as a maintainer

CecileRobertMichon replied:
there's no max number of reviewers, feel free to add as many reviewers as needed.

rohityadavcloud commented on docs/book/src/capi/providers/cloudstack.md:
```diff
@@ -0,0 +1,58 @@
+
+## Hypervisor
+
+The image is built using KVM hypervisor as a `qcow2` image.
```
@Pearl1594 can you help review the technical build steps?

rohityadavcloud commented on docs/book/src/capi/providers/cloudstack.md:
````diff
@@ -0,0 +1,58 @@
+```
+### Prebuilt Images
+
+For convenience, prebuilt images can be found [here](http://download.cloudstack.org/templates/capi/)
````
The link may need review/fixing, as the download.c.o server has limited space, we might move this to a S3 folder as we may want to build/host CAPI templates for KVM/VMware/XenServer/XCP-ng periodically to support various releases of k8s.

davidjumani replied:
Will update this once the download location is confirmed

rohityadavcloud commented on images/capi/Makefile:
```diff
@@ -275,6 +284,7 @@ DO_BUILD_NAMES 			?=	do-centos-7 do-ubuntu-1804 do-ubuntu-2004
 
 QEMU_FLATCAR_BUILD_NAMES	?=	qemu-flatcar
 QEMU_BUILD_NAMES			?=	qemu-ubuntu-1804 qemu-ubuntu-2004 qemu-centos-7 qemu-ubuntu-2004-efi qemu-rhel-8 qemu-rockylinux-8
+CLOUDSTACK_BUILD_NAMES		?=	cloudstack-ubuntu-2004 cloudstack-rhel-8 cloudstack-rockylinux-8
```
We had added support for el7, should we add support for EL7 as well. (CentOS7 is supporte and EOLs in 2024). What about Ubuntu 22.04 due in few weeks?

davidjumani replied:
Added R8 and U20. Will add support as and when new releases come out and are needed

rohityadavcloud commented on images/capi/hack/ensure-vhdutil.sh:
```diff
@@ -0,0 +1,34 @@
+fi
+
+if ! command -v vhd-util >/dev/null 2>&1; then
+    wget http://packages.shapeblue.com.s3-eu-west-1.amazonaws.com/systemvmtemplate/vhd-util -O /tmp/vhd-util
```
@davidjumani do we need vhd-util for convertion, I'm asking as qemu-img convert support a `vpc` output that may be compatible/works with XCP8.x and XS7.x LTS releases.

davidjumani replied:
Have followed https://github.com/apache/cloudstack/blob/main/tools/appliance/build.sh
Tested the converted templates confirmed to work on XS71 and XPC82

davidjumani commented:
/verify-owners

rohityadavcloud commented:
/lgtm
/approve

rohityadavcloud commented:
/assign codenrhoden

rohityadavcloud commented:
/ok-to-test

rohityadavcloud commented:
/retest

davidjumani commented:
/retest

davidjumani commented:
/retest

davidjumani commented:
/retest

codenrhoden commented on images/capi/packer/cloudstack/packer.json:
```diff
@@ -0,0 +1,152 @@
+    "machine_id_mode": "444",
+    "memory": "2048",
+    "output_directory": "./output/{{user `build_name`}}-kube-{{user `kubernetes_semver`}}",
+    "provider": "provider=cloudstack",
```
why not just add this to `ansible_extra_vars`?

davidjumani replied:
Thanks for the feedback, updated!

codenrhoden commented on images/capi/packer/cloudstack/packer.json:
```diff
@@ -0,0 +1,152 @@
+      "vars_inline": {
+        "ARCH": "amd64",
+        "OS": "{{user `distro_name` | lower}}",
+        "PROVIDER": "qemu",
```
Should this be `cloudstack` instead?

davidjumani replied:
Good catch! Updated, thanks

codenrhoden commented on images/capi/packer/cloudstack/packer.json:
```diff
@@ -0,0 +1,152 @@
+      "ssh_password": "{{user `ssh_password`}}",
+      "ssh_timeout": "2h",
+      "ssh_username": "{{user `ssh_username`}}",
+      "type": "qemu",
```
I'm a little worried that this is a 3rd provider that uses the `qemu` builder, but we have separate folders/names for them (`"qemu"`, `"raw"`, and now `"cloudstack"`). There doesn't seem to be much difference between the three of them.

What is making the "cloudstack" builder unique such that it's better as it's own builder rather than a variant/options provided to the "qemu" builder?

davidjumani replied:
Thanks for the feedback @codenrhoden 
CloudStack initially uses the qemu builder to build the capi image. Once built, the post processors kick in to convert them to templates that can be used by the other hypervisors that CloudStack supports, XenServer and VMware
I'm fairly new to packer so do let me know if there's a simpler way to go about this. Thanks :)

codenrhoden commented on images/capi/hack/ensure-vhdutil.sh:
```diff
@@ -0,0 +1,33 @@
+if ! command -v vhd-util >/dev/null 2>&1; then
+    wget http://packages.shapeblue.com.s3-eu-west-1.amazonaws.com/systemvmtemplate/vhd-util
+    chmod +x vhd-util
+    wget http://packages.shapeblue.com.s3-eu-west-1.amazonaws.com/systemvmtemplate/libvhd.so.1.0
```
seeing this made me realize that it's probably worth mentioning in the docs that you can only do this build on a Linux machine (not MacOS or Windows). I know the docs only show examples for Ubuntu and it's implied, but let's make it explicit.

davidjumani replied:
Added a note in the docs, thanks!

codenrhoden reviewed (commented):
Overall I question whether this is really it's own provider or is better as a customization to the existing `qemu` builder.

davidjumani commented:
@codenrhoden /retest

davidjumani commented:
@codenrhoden I've updated the PR to add CloudStack as a subset of qemu rather than have its own directory based on env vars. I've also updated the docs accordingly letting users know which scripts to run to convert the templates. Let me know what you think :)

davidjumani commented:
/retest-required

davidjumani commented on images/capi/ansible/roles/providers/tasks/cloudstack.yml:
```diff
@@ -0,0 +1,72 @@
+  reboot:
+    msg: Ansible is rebooting the VM now
+  when: ansible_os_family == "RedHat" and
+    ansible_kernel != latest_kernel
```
@codenrhoden Moved the os updates to the firstboot role, similar to photon. The VM is then rebooted if cloudstack is the provider and then continues as usual

rohityadavcloud commented:
/lgtm

davidjumani commented:
The repo has been migrated to https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack
@rohityadavcloud @EleanorRigby @CecileRobertMichon @codenrhoden Can you please review. Thanks

rohityadavcloud commented:
/lgtm
/approve

codenrhoden commented on images/capi/ansible/roles/firstboot/tasks/cloudstack.yml:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2020 The Kubernetes Authors.
```
As a brand new file, this should be 2022. The year is always when the file added to the repo.

davidjumani replied:
Thanks for catching this, fixed now

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -100,7 +126,7 @@
     "accelerator": "kvm",
     "ansible_common_vars": "",
-    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3",
+    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3 {{env `ANSIBLE_EXTRA_VARS`}}",
```
I'm not opposed to this, but I am concerned because it makes this one instance of `ansible_extra_vars` unique -- no other providers automatically pull in vars from an environment variable:

```
$ git grep ansible_extra_vars
packer/ami/flatcar.json:  "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
packer/ami/packer-windows.json:        "{{user `ansible_extra_vars`}}",
packer/ami/packer-windows.json:    "ansible_extra_vars": "",
packer/ami/packer.json:        "{{user `ansible_extra_vars`}}",
packer/ami/packer.json:    "ansible_extra_vars": "",
packer/azure/flatcar-gen2.json:  "ansible_extra_vars": "ansible_python_interpreter=/opt/pypy/bin/pypy",
packer/azure/flatcar.json:  "ansible_extra_vars": "ansible_python_interpreter=/opt/pypy/bin/pypy",
packer/azure/packer-windows.json:        "{{user `ansible_extra_vars`}}",
packer/azure/packer-windows.json:    "ansible_extra_vars": "",
packer/azure/packer.json:        "{{user `ansible_extra_vars`}}",
packer/azure/packer.json:        "{{user `ansible_extra_vars`}}",
packer/azure/packer.json:    "ansible_extra_vars": "",
packer/digitalocean/packer.json:        "{{user `ansible_extra_vars`}}",
packer/digitalocean/packer.json:    "ansible_extra_vars": "",
packer/gce/packer.json:        "{{user `ansible_extra_vars`}}"
packer/gce/packer.json:    "ansible_extra_vars": "",
packer/oci/packer.json:        "{{user `ansible_extra_vars`}}",
packer/oci/packer.json:    "ansible_extra_vars": "",
packer/ova/packer-node.json:        "{{user `ansible_extra_vars`}}",
packer/ova/packer-node.json:        "{{user `ansible_extra_vars`}}",
packer/ova/packer-node.json:    "ansible_extra_vars": "guestinfo_datasource_slug={{user `guestinfo_datasource_slug`}} guestinfo_datasource_ref={{user `guestinfo_datasource_ref`}} guestinfo_datasource_script={{user `guestinfo_datasource_script`}}",
packer/ova/packer-windows.json:        "{{user `ansible_extra_vars`}}",
packer/ova/packer-windows.json:    "ansible_extra_vars": "",
packer/qemu/packer.json:        "{{user `ansible_extra_vars`}}"
packer/qemu/packer.json:    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3",
packer/qemu/qemu-centos-7.json:  "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python extra_arguments =vvvv",
packer/qemu/qemu-flatcar.json:  "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
packer/raw/packer.json:        "{{user `ansible_extra_vars`}}"
packer/raw/packer.json:    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3",
packer/raw/raw-flatcar.json:  "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
packer/vbox/packer-windows.json:        "{{user `ansible_extra_vars`}}",
packer/vbox/packer-windows.json:    "ansible_extra_vars": "",
```

davidjumani replied:
Thanks for the review @codenrhoden
Could you let me know a way to pass the provider to ansible without modifying any file ? I'd like the build steps to be quite simple to create a cloudstack capi image

davidjumani replied:
Thanks for the review. I've added the `ansible_user_vars` and added a note in the docs to create an `extra_vars.json` to override it and pass it via `PACKER_VAR_FILES`

codenrhoden reviewed (commented):
I just a have one nit on a copyright header.

Also, while this looks good to go, we can't merge in it's current state due to there being merge commits (see the label). now that this PR is mature, I would recommend you squash the commits. But your call how you want to arrange the commits if you want to have more than one.

rohityadavcloud commented:
/lgtm
/approve

davidjumani commented:
Thanks for your support @codenrhoden @CecileRobertMichon  :)
